Support for Altair

zah opened the pull request:
Includes implementation of sync committees and other Altair-related changes.

arnetheduck commented on beacon_chain/beacon_node_types.nim:
```diff
@@ -155,3 +185,10 @@ type
 func shortLog*(v: AttachedValidator): string = shortLog(v.pubKey)
+
+func hash*(x: SyncCommitteeMsgKey): Hash =
+  result = result !& hash(x.originator.uint64)
```
wow, this is unusually ugly

zah replied:
There is a default implementation for objects doing the same. For some reason, it was not being selected in mysterious way and I didn't want to spend time tracking down the issue (it can be done later).

arnetheduck replied:
yeah, it just amazes me that people like this syntax.

arnetheduck commented on beacon_chain/consensus_object_pools/blockchain_dag.nim:
```diff
@@ -545,6 +548,12 @@ func stateCheckpoint*(bs: BlockSlot): BlockSlot =
     bs = bs.parentOrSlot
   bs
 
+proc forkAtSlot*(dag: ChainDAGRef, slot: Slot): Fork =
```
fork is done at an epoch boundary

tersec commented on tests/simulation/start.sh:
```diff
@@ -205,6 +205,7 @@ GENESIS_DELAY: 10
 GENESIS_FORK_VERSION: 0x00000000
 DEPOSIT_CONTRACT_ADDRESS: ${DEPOSIT_CONTRACT_ADDRESS}
 ETH1_FOLLOW_DISTANCE: 1
+ALTAIR_FORK_EPOCH: 1
```
There's some value in having a couple of real phase 0 epochs, to make sure phase 0 keeps getting tested.  `ALTAIR_FORK_EPOCH` of 1 would mean no real phase 0 epoch processing testing.

zah replied:
I've copied this from the other simulation script. Will fix in both places.

tersec commented on beacon_chain/nimbus_beacon_node.nim:
```diff
@@ -702,20 +713,105 @@ proc subscribeAttestationSubnetHandlers(node: BeaconNode) {.
+
+proc addAltairMessageHandlers(node: BeaconNode, slot: Slot)
+                             {.raises: [Defect, CatchableError].} =
+  node.addPhase0MessageHandlers(node.dag.forkDigests.altair)
```
These aren't exactly phase 0 message handlers at this point, but rather message handlers that exist/overlap for both phase 0 and Altair.

arnetheduck replied:
I imagine we should have a type `Topics = object; blocks: string; aggregates: string etc` or something like that in eth2_network that has an `init` that takes a forkdigest instead of calling the spec function every time with a digest and generating the string anew

arnetheduck replied:
this would make a good `unstable` PR, to intro the handlers in a more clean way

tersec commented on beacon_chain/nimbus_beacon_node.nim:
```diff
@@ -702,20 +713,105 @@ proc subscribeAttestationSubnetHandlers(node: BeaconNode) {.
+  aggregateTopicParams.validateParameters().tryGet()
+  basicParams.validateParameters.tryGet()
+
+proc addPhase0MessageHandlers(node: BeaconNode, forkDigest: ForkDigest)
```
This is a misleading function name.

zah replied:
Admittedly, a bit misleading - the intention was that these are "topics originating in phase0"

arnetheduck commented on beacon_chain/gossip_processing/eth2_processor.nim:
```diff
@@ -345,3 +352,46 @@ proc voluntaryExitValidator*(
+
+  # Potential under/overflows are fine; would just create odd metrics and logs
+  let delay = wallTime - syncCommitteeMsg.slot.toBeaconTime
+  debug "Sync committee message received", delay
```
this needs to shortlog the message

arnetheduck commented on beacon_chain/consensus_object_pools/block_pools_types.nim:
```diff
@@ -192,6 +192,9 @@ type
     shuffled_active_validator_indices*: seq[ValidatorIndex]
+    # TODO this updates at a lower cadence
+    #      (once per EPOCHS_PER_SYNC_COMMITTEE_PERIOD epochs)
+    sync_committee*: seq[ValidatorIndex]
```
how big is the sync committee typically?

zah replied:
It has a fixed size (512). This particular field can be removed now.

arnetheduck commented on beacon_chain/beacon_chain_db.nim:
```diff
@@ -34,6 +34,12 @@ type
+
+  ImmutableValidatorsMemoryStore* = object
+    byIndex*: seq[ImmutableValidatorData2]
+    byPubKey: Table[RawPubKey, int]
```
```suggestion
    byPubKey: Table[ValidatorPubKey, int]
```

zah replied:
I added this type as a work-around for some compilation problems brought by the use of `ValidatorPubKey` as a key. I can revisit them now.

Having that said, the table is not needed anymore, so I'm considering either removing the new abstraction or leaving it in the code in a disabled state that can be easily re-enabled.

arnetheduck replied:
remove - there are local solutions for most problems I know of needing this index that use less memory

zah replied:
What I meant is that I can leave the `ImmutableValidatorsMemoryStore` type in there in order to simplify the life of anyone who wants to revive this functionality.

arnetheduck replied:
still remove :) but less strong opinion

arnetheduck replied:
https://github.com/status-im/nimbus-eth2/pull/2813

arnetheduck commented on beacon_chain/gossip_processing/eth2_processor.nim:
```diff
@@ -20,6 +20,10 @@ import
   ../beacon_node_types,
   ../beacon_clock
 
+from ../spec/datatypes/altair import
```
all of altair already imported

arnetheduck commented on beacon_chain/gossip_processing/eth2_processor.nim:
```diff
@@ -150,9 +157,14 @@ proc blockValidator*(
+  block:
+    when signedBlock is altair.SignedBeaconBlock:
+      logScope:
+        sync_aggregate = $(signedBlock.message.body.sync_aggregate.sync_committee_bits)
```
doesn't `sync_aggregate` have a `shortLog`?

zah replied:
Since we are not logging other block properties, adding this property seems inconsistent. Right now, it serves as a useful debugging aid while the sync committee code is still fresh.

arnetheduck replied:
we log the summarized contents of the block and the root thanks to a logscope above.

perhaps add it to the `shortLog` of the altair blocktype itself then, for example how many bits are set - we log attestation counts and some other "aggregate" data there which is quite useful.

arnetheduck commented on beacon_chain/gossip_processing/eth2_processor.nim:
```diff
@@ -345,3 +357,49 @@ proc voluntaryExitValidator*(
+                                             checkSignature)
+  if v.isErr():
+    debug "Dropping sync committee contribution and proof",
+           validationError = v.error
```
include the message that's being dropped

zah replied:
Do you think this should be done, because the code is new? Dropped attestations are not subjected to the same treatment for example.

arnetheduck replied:
the aim is to log a reasonable amount of content at least _once_ in the processing pipeline, along with a unique identifier (root typically) that can be used to see where in the pipeline the message gets dropped - ie we should have corelated logs at entry and any exit (failed, succeeded etc) - this is done fairly systematically for attestations and blocks

arnetheduck replied:
we also use certain keywords to make grepping easier, ie `Block received` at entry, all important block messages starting with `Block ` etc - these are "soft" rules that make debugging easier thanks to their informal structure

arnetheduck commented on beacon_chain/gossip_processing/gossip_validation.nim:
```diff
@@ -763,3 +763,218 @@ proc validateVoluntaryExit*(
+  # Note this validation implies the validator is part of the broader
+  # current sync committee along with the correct subcommittee.
+  let positionInSubcommittee = dag.getSubcommitteePosition(
+    msg.slot + 1, subnetId, ValidatorIndex msg.validator_index)
```
`validator_index` needs a range check before any cast/conversion, aka never cast / convert externally supplied data

arnetheduck commented on beacon_chain/gossip_processing/gossip_validation.nim:
```diff
@@ -763,3 +763,218 @@ proc validateVoluntaryExit*(
+    # [IGNORE] The contribution's slot is for the current slot
+    # (with a MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance)
+    # i.e. contribution.slot == current_slot.
+    let res = check_propagation_slot_range(msg.message.contribution.slot, wallTime)
```
`?`?

arnetheduck commented on beacon_chain/consensus_object_pools/blockchain_dag.nim:
```diff
@@ -956,6 +977,106 @@ proc pruneBlocksDAG(dag: ChainDAGRef) =
+
+  @(toOpenArray(syncCommittee, startIdx, onePastEndIdx - 1))
+
+const emptySyncCommittee = newSeq[ValidatorPubKey]()
```
why the const? `@[]` should be enough?

zah replied:
Some of the functions (templates) were supposed to return openarray values in order to avoid allocations. This const was supposed to cover the case where you have an empty openarray result. The whole plan didn't work out unfortunately due to the limited support for composing `openarray` expressions in Nim 1.2.

arnetheduck replied:
ok, let's clean it up? I'm somewhat wary of pre-optimizing the code for 1.6 when 1.6 isn't ready yet and won't be for a good while - this may or may not turn out to be a real problem and it looks off compared to the rest of th ecode

tersec replied:
I removed `emptySyncCommittee` for now -- if it's useful, it can be re-added easily enough.

arnetheduck commented on beacon_chain/consensus_object_pools/sync_committee_msg_pool.nim:
```diff
@@ -0,0 +1,167 @@
+{.push raises: [Defect].}
+
+import
+  sets,
```
```suggestion
  std/sets,
```

arnetheduck commented on beacon_chain/consensus_object_pools/sync_committee_msg_pool.nim:
```diff
@@ -0,0 +1,167 @@
+      BestSyncSubcommitteeContribution(
+        totalParticipants: totalParticipants,
+        participationBits: contribution.aggregation_bits,
+        signature: contribution.signature.load.get)
```
if we're assuming only loadable signatures get this far, shouldn't we use a type to enforce it?

arnetheduck commented on beacon_chain/consensus_object_pools/sync_committee_msg_pool.nim:
```diff
@@ -0,0 +1,167 @@
+    except KeyError:
+      raiseAssert "We have checked for the key upfront"
+  else:
+    SyncAggregate(sync_committee_signature: ValidatorSig.infinity)
```
`SyncAggregate.init()` which does the infinity sig thing?

arnetheduck commented on beacon_chain/consensus_object_pools/sync_committee_msg_pool.nim:
```diff
@@ -0,0 +1,167 @@
+        let globalPos = subnetId * SYNC_SUBCOMMITTEE_SIZE + pos
+        result.sync_committee_bits.setBit globalPos
+
+    if not initialized:
```
this code would be simplified by simply starting from an infinity point and adding every signature .. costly? easy? ping @mratsim

mratsim replied:
Cheap, or use "aggregateAll"
 https://github.com/status-im/nim-blscurve/blob/8cbc4ae8709b94c026a015dd03933e5bbe8a0040/blscurve/blst/blst_min_pubkey_sig_core.nim#L183

arnetheduck replied:
hm, that function looks like it should set to infinity on "no entries"

arnetheduck commented on beacon_chain/gossip_processing/gossip_validation.nim:
```diff
@@ -763,3 +763,218 @@ proc validateVoluntaryExit*(
+
+    for validatorPubKey in dag.syncCommitteeParticipants(
+        msg.message.contribution.slot + 1,
+        SubnetId msg.message.contribution.subcommittee_index,
```
range check

zah replied:
Already done above:

```nim
  block:
    # [REJECT] The subcommittee index is in the allowed range
    # i.e. contribution.subcommittee_index < SYNC_COMMITTEE_SUBNET_COUNT.
    if msg.message.contribution.subcommittee_index >= SYNC_COMMITTEE_SUBNET_COUNT:
      return err((ValidationResult.Reject, cstring(
        "validateSignedContributionAndProof: subcommittee index too high")))
```

arnetheduck replied:
right

arnetheduck commented on beacon_chain/networking/network_metadata.nim:
```diff
@@ -162,7 +162,8 @@ const
   pyrmontMetadata* = eth2Network "shared/pyrmont"
   praterMetadata* = eth2Network "shared/prater"
-  altairDevnet0Metadata* = eth2Network "shared/altair-devnet-0"
+  altairDevnet1Metadata* = eth2Network "shared/altair-devnet-1"
```
obsolete

arnetheduck commented on beacon_chain/nimbus_beacon_node.nim:
```diff
@@ -38,17 +38,23 @@ import
 
 from eth/common/eth_types import BlockHashOrNumber
 
+from spec/datatypes/altair import
```
already imported

arnetheduck commented on beacon_chain/spec/forkedbeaconstate_helpers.nim:
```diff
@@ -278,3 +278,22 @@ func shortLog*(x: ForkedSignedBeaconBlock | ForkedTrustedSignedBeaconBlock): aut
 chronicles.formatIt ForkedSignedBeaconBlock: it.shortLog
 chronicles.formatIt ForkedTrustedSignedBeaconBlock: it.shortLog
+
+proc forkAtEpoch*(cfg: RuntimeConfig, epoch: Epoch): Fork =
```
move these to `beaconstate.nim`?

arnetheduck replied:
stays here after rename

arnetheduck commented on beacon_chain/spec/forkedbeaconstate_helpers.nim:
```diff
@@ -278,3 +278,22 @@ func shortLog*(x: ForkedSignedBeaconBlock | ForkedTrustedSignedBeaconBlock): aut
+  else:
+    altairFork(cfg)
+
+proc forkVersionAtSlot*(cfg: RuntimeConfig, slot: Slot): Version =
```
`...AtEpoch`

arnetheduck commented on beacon_chain/spec/state_transition.nim:
```diff
@@ -402,6 +402,7 @@ proc makeBeaconBlock*(
     proposerSlashings: seq[ProposerSlashing],
     attesterSlashings: seq[AttesterSlashing],
     voluntaryExits: seq[SignedVoluntaryExit],
+    sync_aggregate: SyncAggregate,
```
unused parameter

arnetheduck commented on beacon_chain/consensus_object_pools/blockchain_dag.nim:
```diff
@@ -1209,3 +1309,21 @@ proc getProposer*(
 
   proposer
+
+template validateValidatorIndexOr*(idxParam: uint64, dag: ChainDAGRef,
```
Hm, the notion of a "valid" `ValidatorIndex` could be more accurate if we define it as being in the range `[0, VALIDATOR_REGISTRY_LIMIT)` - this also allows it to capure "future" validators that are in the process of being added - it's also roughly what we want to capture in the API parts of the code.

Being valid against our immutable validator cache length is a bit of a strange in-between that relies on a second-degree guarantee that the cache will get updated in a particular way - the cache has its own lifetime and though it generally lines up against the places where validator indices are used, but it's not 1:1 match either - `CachedValidatorIndex` perhaps would be a good way to signal such validity, so that it doesn't end up getting used in the wrong way - we could put this type in chandag somewhere as well, because that's roughly the scope in which it is valid.

arnetheduck replied:
The problem in particular happens when we're processing state transitions - although we're passing a `ValidatorIndex` that is valid in _some_ history, it doesn't mean that it's valid in that particular state - it also extends to cases such as `EpochRef` and the balances cache which also doesn't follow the cache validity.

zah replied:
I agree with the arguments, but I think this PR carries enough `ValidatorIndex` reforms already. More granularity in the semantics of these types can be delivered in future PRs.

arnetheduck replied:
I suspect what I describe might be closer to the current setup and would be less changes "total" because the code was already heading in that direction.

btw a similar situation exists with committeeindex which is statically limited to the range `[0, MAX_COMMITTEES_PER_SLOT)`.

arnetheduck replied:
Moving discussion to https://github.com/status-im/nimbus-eth2/pull/2813

arnetheduck commented on beacon_chain/validators/validator_duties.nim:
```diff
@@ -171,9 +169,11 @@ proc sendAttestation*(
 
   return case ok
     of ValidationResult.Accept:
+      # TODO altair: this probably is just a function of wallslot's altair, not
```
investigate this? seems important to not lose attestations during transition - however, clients are listening to both topics for an overlap period, right?

tersec replied:
Sure, attestations remain unchanged, so it doesn't really matter which forkdigest-topic it gets sent to. https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-beta.2/specs/altair/p2p-interface.md#transitioning-the-gossip states

> A node's behavior before the fork and after the fork are as follows:
> Pre-fork:
> - Peers who propagate messages on the post-fork topics MAY be scored negatively proportionally to time till fork,
  to account for clock discrepancy.
> - Messages can be IGNORED on the post-fork topics, with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` margin.
>
> Post-fork:
> - Peers who propagate messages on the pre-fork topics MUST NOT be scored negatively. Lagging IWANT may force them to.
> - Messages on pre and post-fork variants of topics share application-level caches.
>  E.g. an attestation on the both the old and new topic is ignored like any duplicate.
> - Two epochs after the fork, pre-fork topics SHOULD be unsubscribed from. This is well after the configured `seen_ttl`.

Which this seems broadly compatible with.

It still looks like the wall slot/epoch, as this comment suggests, is correct, but I can't find specifically spec text to support this.

tersec replied:
https://github.com/status-im/nimbus-eth2/pull/2773 switches this to wall time

arnetheduck commented on beacon_chain/spec/signatures_batch.nim:
```diff
@@ -128,6 +129,8 @@ proc addAttestation(
       attestation: Attestation,
       validatorKeys: auto,
       state: ForkedHashedBeaconState,
+      fork: Fork,
```
attestations (and all other block fields) must use the fork of the attestation, not that of the block

arnetheduck replied:
removed these changes while rebasing to return to the previous setup which worked

arnetheduck commented on beacon_chain/nimbus_beacon_node.nim:
```diff
@@ -1024,6 +1054,11 @@ proc onSlotEnd(node: BeaconNode, slot: Slot) {.async.} =
     next_action_wait.set(nextActionWaitTime.toFloatSeconds)
 
+  let epoch = slot.epoch
+  if epoch >= node.network.forkId.next_fork_epoch:
```
move the conditional to `eth2_network` as well?

arnetheduck commented on beacon_chain/consensus_object_pools/blockchain_dag.nim:
```diff
@@ -980,6 +982,104 @@ proc pruneBlocksDAG(dag: ChainDAGRef) =
+    return
+
+  let
+    headSlot = dag.headState.data.hbsAltair.data.slot
```
not sure it matters here, but when there block goes missing in a slot, this head slot does not advance

cheatfate commented on beacon_chain/rpc/rest_validator_api.nim:
```diff
@@ -35,16 +36,10 @@ proc installValidatorApiHandlers*(router: var RestRouter, node: BeaconNode) =
-              return RestApiResponse.jsonError(Http500,
-                                            UnsupportedValidatorIndexValueError)
-          res.add(vres.get())
+          let idx = uint64(item).validateValidatorIndexOr(node.dag):
```
This code is incorrect, old code acts properly and returns proper error codes.
1. If `validator_index` is in `[0, 2^40)` for 64bit systems, it is valid `validator_index` and should be used for query.
2. If we have 32bit version we allow `validator_index` only in range of `[0, 2^32)`, if validator_index is in range of `[2^32, 2^40)` we return error 500 because we do not support this values on 32bit (Nim array indexes are 32bit).

So you should not generated error `400` if `validator_index` is valid value, if its exceeds current validators number you should just return an empty response with status code `200`.

https://ethereum.github.io/eth2.0-APIs/#/Validator/getAttesterDuties

tersec commented on beacon_chain/gossip_processing/gossip_validation.nim:
```diff
@@ -468,10 +473,10 @@ proc validateAggregate*(
     of BatchResult.Timeout:
       beacon_aggregates_dropped_queue_full.inc()
-      return err((ValidationResult.Reject, cstring("validateAggregate: timeout checking slot signature")))
+      return errReject("validateAggregate: timeout checking slot signature")
```
Shouldn't `errReject` on timeouts.

tersec commented on beacon_chain/gossip_processing/gossip_validation.nim:
```diff
@@ -493,11 +497,10 @@ proc validateAggregate*(
     of BatchResult.Timeout:
       beacon_aggregates_dropped_queue_full.inc()
-      return err((ValidationResult.Reject, cstring("validateAggregate: timeout checking aggregate signature")))
+      return errReject("validateAggregate: timeout checking aggregate signature")
```
Shouldn't `errReject` on timeouts.

tersec commented on beacon_chain/gossip_processing/gossip_validation.nim:
```diff
@@ -480,11 +485,10 @@ proc validateAggregate*(
     of BatchResult.Timeout:
       beacon_aggregates_dropped_queue_full.inc()
-      return err((ValidationResult.Reject, cstring("validateAggregate: timeout checking aggregator signature")))
+      return errReject("validateAggregate: timeout checking aggregator signature")
```
Shouldn't `errReject` on timeouts.

arnetheduck commented on tests/test_macros.nim:
```diff
@@ -0,0 +1,17 @@
+import
```
testing leftover?

arnetheduck commented on beacon_chain/beacon_chain_db.nim:
```diff
@@ -34,6 +34,10 @@ type
 
+  ImmutableValidatorsMemoryStore* = object
+    byIndex*: seq[ImmutableValidatorData2]
+    byPubKey: Table[ValidatorPubKey, ValidatorIndex]
```
so.. remove?

arnetheduck replied:
(yagni)

tersec commented on beacon_chain/rpc/rest_utils.nim:
```diff
@@ -150,6 +150,11 @@ func match(data: openarray[char], charset: set[char]): int =
       return 1
   0
 
+proc `==`*(lhs, rhs: RestValidatorIndex): bool {.borrow, noSideEffect.}
```
maybe use `template`/`distinctBase`

tersec commented on beacon_chain/consensus_object_pools/sync_committee_msg_pool.nim:
```diff
@@ -0,0 +1,162 @@
+      raiseAssert "We have checked for the key upfront"
+  else:
+    SyncAggregate(sync_committee_signature: ValidatorSig.infinity)
+
```
trailing newline

arnetheduck committed:
Initial implementation of sync committees and altair block production

arnetheduck committed:
Add server getBlockV2(), and produceBlockV2().

Add getBlockV2() to REST test suite.
Add client getBlockV2(), and produceBlockV2().
Fix URLs in comments.
Add some primitives and fix some issues in forks.nim.
Switch `validator_client` to V2 calls usage.
Bump `chronos` with imports fixes.
Bump `nim-json-serialization` for `requireAllFields`.

tersec commented on beacon_chain/validators/validator_duties.nim:
```diff
@@ -517,6 +584,147 @@ proc handleAttestations(node: BeaconNode, head: BlockRef, slot: Slot) =
+
+  for committeeIdx in allSyncCommittees():
+    for valKey in syncSubcommittee(syncCommittee, committeeIdx):
+      let validator = node.getAttachedValidator(valKey)
```
Running `node.getAttachedValidator()` for many validators can get quite expensive, each being an O(n) linear search dependent on total validators, but this is something that can be adjusted after merging.

zah replied:
`node.getAttachedValidator(pubKey)` is searching in hash table stored in the validator pool.

tersec commented on beacon_chain/validators/validator_duties.nim:
```diff
@@ -517,6 +584,147 @@ proc handleAttestations(node: BeaconNode, head: BlockRef, slot: Slot) =
+      # TODO Hoist outside of the loop with a view type
+      #      to avoid the repeated offset calculations
+      for valKey in syncSubcommittee(syncCommittee, committeeIdx):
+        let validator = node.getAttachedValidator(valKey)
```
It might be useful to construct a lookup table here to only scan this once, but in the absence of good tests for correctness of such, it's something to do after the `unstable` merge for risk mitigation purposes.

If it's a problem empirically, it'll show up soon enough, or would already have done so on the Pyrmont testnet.

tersec commented on beacon_chain/validators/validator_duties.nim:
```diff
@@ -781,6 +990,10 @@ proc handleValidatorDuties*(node: BeaconNode, lastSlot, slot: Slot) {.async.} =
     let sendAggregatedAttestationsFut =
       sendAggregatedAttestations(node, head, slot)
 
+    let handleSyncCommitteeContributionsFut =
```
This sends contributions even when operating in phase 0 mode, e.g.,
> "lvl":"NOT","ts":"2021-08-29 11:52:15.010+02:00","msg":"Contributions sent","topics":"beacval","tid":579806,"file":"validator_duties.nim:659","count":0,"time":0

Should be somewhat harmless, though.

zah replied:
The log messages has been silenced in the `count = 0` case in https://github.com/status-im/nimbus-eth2/pull/2836.
To be clear, the code never sends anything on the wire before the Altair transition epoch. This happens to be the case because all sync committees are effectively empty prior to the transition.
